packages/python/plotly/plotly/express/_core.py

```diff
@@ -1607,39 +1774,70 @@ def build_dataframe(args, constructor):
             args["histfunc"] = None
             args["orientation"] = "h"
             args["x"] = count_name
-            df_output[count_name] = 1
+            df_output = df_output.with_columns(**{count_name: nw.lit(1)})
         else:
             args["histfunc"] = None
             args["orientation"] = "v"
             args["y"] = count_name
-            df_output[count_name] = 1
+            df_output = df_output.with_columns(**{count_name: nw.lit(1)})
 
     if no_color:
         args["color"] = None
     args["data_frame"] = df_output
     return args
 
 
-def _check_dataframe_all_leaves(df):
-    df_sorted = df.sort_values(by=list(df.columns))
-    null_mask = df_sorted.isnull()
-    df_sorted = df_sorted.astype(str)
-    null_indices = np.nonzero(null_mask.any(axis=1).values)[0]
-    for null_row_index in null_indices:
-        row = null_mask.iloc[null_row_index]
-        i = np.nonzero(row.values)[0][0]
-        if not row[i:].all():
+def _check_dataframe_all_leaves(df: nw.DataFrame) -> None:
+    cols = df.columns
+    df_sorted = df.sort(by=cols, descending=False, nulls_last=True)
+    null_mask = df_sorted.select(*[nw.col(c).is_null() for c in cols])
+    df_sorted = df_sorted.with_columns(nw.col(*cols).cast(nw.String()))
```

Comment:
This syntax is pretty arcane, would be nice to have a more readable syntax for casting column types. (Doesn't need to be changed now, just a feature request!)

Reply:
We have a [PR open for some time](https://github.com/narwhals-dev/narwhals/pull/1045), but unsure if we are going to merge it. Would you say that the following is more readable?

```py
df_sorted = df_sorted.cast({c: nw.String() for c in cols})
```

Reply:
Hmm, that actually feels odd to me in a different way. I would expect something more like simply
```python
df_sorted = df_sorted.cast(nw.String)
```

Reply:
oh sure! such PR would allow that as well. It has gone a bit out of date though. If there is interest I can come back to it

Reply:
Would it work (and be more readable) to rewrite this as:
```python
    null_mask = df_sorted.select(nw.all().is_null())
    df_sorted = df_sorted.select(nw.all().cast(nw.String()))
    null_indices_mask = null_mask.select(
        null_mask=nw.any_horizontal(nw.all())
    ).get_column("null_mask")
```
?

My hesitation with `DataFrame.cast` is that I'd really like to nudge people towards learning about the building blocks like expressions - this does come with a little learning curve, but it does mean that, once users are past that learning curve, they already know how to compose these building blocks when they need to accomplish non-standard tasks